src/buffer.rs

```diff
@@ -921,11 +921,60 @@ where
         self.get_pixel_mut(x, y).blend(&p)
     }
 
+    fn copy_within(&mut self, from: (u32, u32), to: (u32, u32), width: u32, height: u32) -> bool {
```

Comment:
Bikeshedding time. That's a lot of very similar parameters that could easily be switched up by accident. I kind of like the `Rect { }` idea most. The reasons are:

* It is consistent with the two-parameter style of the standard library.
* It groups source selection and destination point which makes it harder to swap them.
* `Rect` can be reused (e.g. [`ImageDecoder::read_rect`](https://docs.rs/image/0.23.0-preview.0/image/trait.ImageDecoderExt.html#method.read_rect), [`sub_image`](https://docs.rs/image/0.23.0-preview.0/image/struct.SubImage.html#method.sub_image))
* A separate struct such as `Rect` offers more extension points for named constructors to further improve clarity but also to add diversity.

Most similar might be the `trait` route, mimicking `slice::copy_within` even more closely. The `Rect` could be simply one type that implements the trait in practice. It might even be that the trait is `Into<Rect>` (although that feels too much like mixing concerns).

Reply:
After a bit of thinking I also prefer the `Rect { }` version the most, while it does require one to type a bit more. It seems to have the most advantages compared to the alternatives. You already gave a lot of good reasons like reuseability and readability. For the other two options, which are both really similar, the only thing I see about them is that they aren't really intuitive making it easy to mix them up. You can only guess what tuple means what and are most likely to get it wrong if you don't look at the parameter names.

Reply:
Do you have the time to implement this version? There does not seem to be any reason not to go with this option then.

Reply:
I can implement it later today, two small questions regarding this then. Where should I put the `Rect` definition and should the target point by a tuple or two seperate args as in:
`image.copy_within(Rect { x: 0, y: 0, w: 20, h: 13) }, (1, 1));`
vs
`image.copy_within(Rect { x: 0, y: 0, w: 20, h: 13) }, 1, 1);`
?

Reply:
Both the pixel related functions as well as operations such as `overlay` take two arguments instead of a tuple to represent coordinates. Maybe it is worth doing an overhaul at some point and unify this to a newtype but it seems more consistent to use two arguments right now.

To me there seem to be two sensible choices to place the implementation:
* in `src/image.rs` and export it at the top level
* in `src/math/rect.rs` and export it in `image::math`

Reply:
A potential counter argument to having the target point as two arguments: It does not mirror `slice::copy_within` as closely. I'd say that it would still be sufficiently similar but we should consider it. Note that the same is true for indexing and the missing `Coord` type is a reason why we do have `ops::Index` implementations... But let's not get ahead of ourselves in this PR.

Reply:
I went with the two arguments approach given most functions use this too as said. Seems better to be uniform for this in my eyes. Put the `Rect` struct into `src/math/rect.rs` since it can, and probably will at some point, be used for more things than just this operation.